src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java

```diff
@@ -220,10 +247,68 @@ public void buildGraph(Graph graph, GraphBuilderModuleSummary graphBuilderModule
             "calculate missing elevations"
         );
         log.info(missingElevationsTask.start());
-        assignMissingElevations(graph, edgesWithElevation);
+        assignMissingElevations(graph, edgesWithCalculatedElevations);
         log.info(missingElevationsTask.finish());
     }
 
+    /**
+     * A runnable that contains the relevant info for executing a process edge operation in a particular thread.
+     */
+    private class ProcessEdgeTask implements Runnable {
+        private final StreetWithElevationEdge swee;
+        private final ConcurrentHashMap<Long, Coverage> coveragesForThread;
+
+        public ProcessEdgeTask(StreetWithElevationEdge swee, ConcurrentHashMap<Long, Coverage> coveragesForThread) {
+            this.swee = swee;
+            this.coveragesForThread = coveragesForThread;
+        }
+
+        @Override public void run() {
+            processEdge(swee, getThreadSpecificCoverageInstance());
+        }
+
+        /**
+         * Get the thread-specific Coverage instance to avoid multiple threads waiting for a lock to be released on a
+         * Interpolator2D instance.
+         */
+        private Coverage getThreadSpecificCoverageInstance () {
```

Comment:
It appears that this method can take some time when it first generates the cache. Are we sure the time to build the coverage for each thread doesn't outweigh the time that deadlocks occur when multiple threads are operating on a single coverage instance? Just curious about hard numbers for a specific graph.

Reply:
> Are we sure the time to build the coverage for each thread doesn't outweigh the time that deadlocks occur when multiple threads are operating on a single coverage instance?

I think you mean to say `...outweigh the time incurred while waiting for locks to be released when...`. A deadlock is when all threads cease operation because they are waiting for interdependent locks in different threads to be released that never will be.

This method definitely speeds things up. [Gard from Entur shared that this improved the Oslo graph build times as follows](https://github.com/opentripplanner/OpenTripPlanner/pull/2990#issuecomment-596974429):

| scenario | time |
|--|--| 
| dev-1.x | 8.7 minutes |
| dev-1x-elevation-improvements (single coverage instance) | 11.4 minutes |
| dev-1x-elevation-improvements (multiple coverage instance) | 5.9 minutes |